plan/gen_physical_plans.go

```diff
@@ -308,67 +352,133 @@ func (p *LogicalJoin) tryToGetIndexJoin() ([]PhysicalPlan, bool) {
 	return plans, false
 }
 
-func (p *LogicalJoin) generatePhysicalPlans() []PhysicalPlan {
-	mergeJoins := p.getMergeJoin()
+// LogicalJoin can generates hash join, index join and sort merge join.
+// Firstly we check the hint, if hint is figured by user, we force to choose the corresponding physical plan.
+// If the hint is not matched, it will get other candidates.
+// If the hint is not figured, we will pick all candidates.
+func (p *LogicalJoin) genPhysPlansByReqProp(prop *requiredProp) []PhysicalPlan {
+	mergeJoins := p.getMergeJoin(prop)
 	if (p.preferJoinType&preferMergeJoin) > 0 && len(mergeJoins) > 0 {
 		return mergeJoins
 	}
 	joins := make([]PhysicalPlan, 0, 5)
 	joins = append(joins, mergeJoins...)
 
-	indexJoins, forced := p.tryToGetIndexJoin()
+	indexJoins, forced := p.tryToGetIndexJoin(prop)
 	if forced {
 		return indexJoins
 	}
 	joins = append(joins, indexJoins...)
 
-	hashJoins := p.getHashJoins()
+	hashJoins := p.getHashJoins(prop)
 	if (p.preferJoinType & preferHashJoin) > 0 {
 		return hashJoins
 	}
 	joins = append(joins, hashJoins...)
 	return joins
 }
 
-func (p *LogicalProjection) generatePhysicalPlans() []PhysicalPlan {
+// tryToGetChildProp will check if this sort property can be pushed or not.
+// When a sort column will be replaced by scalar function, we refuse it.
+// When a sort column will be replaced by a constant, we just remove it.
+func (p *LogicalProjection) tryToGetChildProp(prop *requiredProp) (*requiredProp, bool) {
```

Comment:
The function name is not proper.

Reply:
The returning value is used for child's req prop.